_posts/2024-10-25-topn.md

```diff
@@ -0,0 +1,466 @@
+GROUP BY
+    l_suppkey;
+```
+
+| l_suppkey |        recent_orders        |
+|----------:|-----------------------------|
+| 2992      | [233573, 3597639, 3060227]  |
+| 8516      | [4675968, 5431174, 4626530] |
+| 3205      | [3844610, 4396966, 3405255] |
+| 2152      | [1672000, 4209601, 3831138] |
+| 1880      | [4852999, 2863747, 1650084] |
+| ...       | ...                         |
+
+The `max_by` function is an aggregate function, so it takes advantage of DuckDB's fast hash aggregation rather than sorting.
+Instead of sorting by `l_shipdate`, the `max_by` function scans through the dataset just once and keeps track of the `N` highest `l_shipdate` values.
+It then returns the order id that corresponds with each of the most recent shipment dates.
+The radix sort in DuckDB must scan through the dataset once per byte, so scanning only once providing a significant speedup.
+For example, if sorting by a 64-bit integer, the sort algorithm must loop through the dataset 8 times vs. 1 with this approach!
+A simple benchmark is included below.
+
+However, this SQL query has a few gaps.
+The query returns results as a `LIST` rather than as separate rows.
+Thankfully the `unnest` function can split a `LIST` into separate rows:
+
+```sql
+FROM lineitem
+SELECT 
+    l_suppkey,
+    unnest(
+        max_by(l_orderkey, l_shipdate, 3)
+    ) AS recent_orders
+GROUP BY
+    l_suppkey;
+```
+
+| l_suppkey | recent_orders |
+|----------:|--------------:|
+| 2576      | 930468        |
+| 2576      | 2248354       |
+| 2576      | 3640711       |
+| 5559      | 4022148       |
+| 5559      | 1675680       |
+| 5559      | 4976259       |
+| ...       | ...           |
+
+The next gap is that there is no way to easily see the `l_shipdate` associated with the returned `l_orderkey` values.
+This query only returns a single column, while typically a top N by group analysis will require the entire row.
+
+Fortunately, DuckDB allows us to refer to the entire contents of a row as if it were just a single column!
+By referring to the name of the table itself (here, `lineitem`) instead of the name of a column, the `max_by` function can retrieve all columns.
+
+```sql
+FROM lineitem
+SELECT 
+    l_suppkey,
+    unnest(
+        max_by(lineitem, l_shipdate, 3)
+    ) AS recent_orders
+GROUP BY
+    l_suppkey;
+```
+
+| l_suppkey | recent_orders |
+|----------:|------------------------------------------------------------------------|
+| 5411      | {'l_orderkey': 2543618, 'l_partkey': 105410, 'l_suppkey': 5411, ... |
+| 5411      | {'l_orderkey': 580547, 'l_partkey': 130384, 'l_suppkey': 5411, ...  |
+| 5411      | {'l_orderkey': 3908642, 'l_partkey': 132897, 'l_suppkey': 5411, ... |
+| 90        | {'l_orderkey': 4529697, 'l_partkey': 122553, 'l_suppkey': 90, ... |
+| 90        | {'l_orderkey': 4473346, 'l_partkey': 160089, 'l_suppkey': 90, ... |
+| ...       | ...           |
+
+Well that isn't so friendly looking!
```

Comment:
Why not?

Reply:
Oh, just that the struct is all in 1 column and long. Happy to remove it you would like! Or maybe:
"Let's make this friendlier by splitting the struct back out into separate columns!"

Reply:
Edited! Let me know what you think.